packages/virtual/src/index.tsx

```diff
@@ -1,71 +1,109 @@
 import { For, createSignal } from "solid-js";
-import type { Accessor, JSX } from "solid-js";
+import type { Accessor, JSX, Signal } from "solid-js";
 
 /**
- * A basic virtualized list (see https://www.patterns.dev/vanilla/virtual-lists/) component for improving performance when rendering very large lists
+ * A headless virtualized list (see https://www.patterns.dev/vanilla/virtual-lists/) utility for constructing your own virtualized list components with maximum flexibility.
+ *
+ * @param rootElement accessor for the element to use as the root of the virtualized list
+ * @param items the list of items
+ * @param rootHeight the height of the root element of the virtualizedList
+ * @param rowHeight the height of individual rows in the virtualizedList
+ * @param overscanCount the number of elements to render both before and after the visible section of the list, so passing 5 will render 5 items before the list, and 5 items after. Defaults to 1, cannot be set to zero. This is necessary to hide the blank space around list items when scrolling
+ * @returns an object whose properties are used by the list's components
+ */
+export function createVirtualList<T extends readonly any[]>({
+  rootElement,
+  items,
+  rootHeight,
+  rowHeight,
+  overscanCount,
+}: {
+  rootElement: Accessor<Element>;
+  items: T | undefined | null | false;
+  rootHeight: number;
+  rowHeight: number;
+  overscanCount?: number;
+}): {
+  onScroll: VoidFunction;
+  containerHeight: () => number;
+  viewerTop: () => number;
+  visibleItems: () => readonly T[];
+} {
+  items = items || ([] as any as T);
+  overscanCount = overscanCount || 1;
+
+  const [offset, setOffset] = createSignal(0);
+
+  const getFirstIdx = () => Math.max(0, Math.floor(offset() / rowHeight) - overscanCount);
+
+  const getLastIdx = () =>
+    Math.min(
+      items.length,
+      Math.floor(offset() / rowHeight) + Math.ceil(rootHeight / rowHeight) + overscanCount,
+    );
+
+  return {
+    onScroll: () => {
+      setOffset(rootElement().scrollTop);
+    },
+    containerHeight: () => items.length * rowHeight,
+    viewerTop: () => getFirstIdx() * rowHeight,
+    visibleItems: () => items.slice(getFirstIdx(), getLastIdx()),
```

Comment:
`getFirstIdx` and `getLastIdx` could be returned as well

Reply:
Why's that desirable?

Reply:
`getFirstIdx` could be used to implement a ledger effect.

Reply:
> Personal opinion: A single object returned would make it cleaner. `const myVirtual = createVirtualList(...)` and then `myVirtual.onScroll` can be assigned to elements as necessary. Renaming it is still possible during destructuring, like: `const {onScroll: myOnScroll} = createVirtualList(...)`

yeah i'm on the fence about this one as well

the object could also be made of getters instead of accessors
`virtual.visibleItems` vs `virtual.visibleItems()`
since I doubt anyone will try to destructure it further

Reply:
the [values, setter] thing is nice and symmetrical with the createSignal API, but maybe it's a little too cute. I'm fine with implementing whatever API you settle on out of the options raised in this PR.